src/injectSheet.js

```diff
@@ -1,62 +1,56 @@
 import defaultInjectSheet from 'react-jss'
 
-const managers = new WeakMap()
-
 if (process.env.NODE_ENV === 'production') {
   // eslint-disable-next-line no-console
   console.error('react-jss-hmr should never be used in production!')
 }
 
+const managers = new WeakMap()
+const deferredState = new WeakMap()
+
 export default function injectSheet(...rest) {
   const createHoc = defaultInjectSheet.apply(this, rest)
   return (InnerComponent) => {
     const Jss = createHoc(InnerComponent)
 
     class HotJss extends Jss {
+      constructor(props, context) {
+        super(props, context)
+        deferredState.set(this.manager, this.state)
+      }
+
       componentWillMount() {
-        // Note: this will never be called during hot module replacement, so we can
-        // use it as a place to store away the SheetManager
+        // This will never be called during hot module replacement, so we can use it as a place to
+        // store away the SheetManager
         super.componentWillMount()
         managers.set(Object.getPrototypeOf(this), this.manager)
```

Comment:
Can't this also be in the constructor? Or is the constructor when on HMR?

Reply:
No because the `this` in the constructor is not the proxy.
However I had a read of Redux’ `connectAdvanced()` that you referred to before and if I use their idea - a counter (`version`) - as the key then I could move this into the constructor.

I was thinking there wasn’t much reason to switch to the `version` idea because we need to track the `SheetsManager` anyway, but moving this code to the constructor might be a good enough reason. What do you think?